Support for Configurable Lints

christopher-henderson opened the pull request:
Howdy folks,

The following is an implementation of the notion of configurable lints that we have been working on.

First of all, there two fortunate attributes to this code review...

1. All of the documentation written up in #624 still apply, so if there are any questions on how this behaves or is used, then that document is still accurate.
2. The _wide_ majority of this PR are tests and docs (over 75%).

However, it still may be conducive to provide a walkthrough (or, at least, give an ordering of the file that should be read carefully). I'll endeavor to at least provide said ordering.

1. [v3/lint/configuration.go](https://github.com/zmap/zlint/pull/648/files#diff-ef5cbe9f39c9fefd0fb1ea33b4450798e887987247717128374187802544e13e) - This is the most involved file in the PR. Most others are plumbing.
2. [v3/lint/global_configurations.go](https://github.com/zmap/zlint/pull/648/files#diff-943896debd9f4e05b78cc4761938c132ba6641d8f4d6886e99de28d6d267ac87) - This is a list of the pre-defined configuration sections that ZLint will ship with. Of course they are all empty at the moment, but I think it still bares taking a look at.
3. [CONTRIBUTING.md](https://github.com/zmap/zlint/pull/648/files#diff-eca12c0a30e25b4b46522ebf89465a03ba72a03f540796c979137931d8f92055) - Docs on how to work with this system from a lint are important.

The rest are miscellaneous plumbing required to actually hookup the feature end-to-end as well as enabling configurations from within unit tests where desired.

This PR resolves #624

christopher-henderson commented on v3/cmd/zlint/main.go:
```diff
@@ -47,6 +47,8 @@ var ( // flags
 	includeSources  string
 	excludeSources  string
 	printVersion    bool
+	config          string
```
The changes in `main` are to:

1. Add the new flags that either point to a configuration file are prints an example default one to the terminal.
2. If printing the default example is selected, then that is done and the problem exists. Any other flag is ignored. This behavior is consistent with printing a list of all available lints as well as printing ZLint's version.

christopher-henderson commented on v3/lint/configuration.go:
```diff
@@ -0,0 +1,227 @@
+package lint
```
This file is the meat of the change.

christopher-henderson commented on v3/lint/configuration.go:
```diff
@@ -0,0 +1,242 @@
+	return c.resolveHigherScopedReferences(target)
+}
+
+// resolveHigherScopeReferences takes in an interface{} value and attempts to
```
Indeed, this method is what accomplished the somewhat fancy scoping feature by reflecting into the struct provided by the lint and trying to find and data member that is one of the higher scoped configuration fields.

christopher-henderson commented on v3/lint/configuration_test.go:
```diff
@@ -0,0 +1,913 @@
+package lint
```
These tests require custom structs for each case that we would like to exercise. That, coupled with the fact the types are not values in Go, means that table driven testing cannot reasonably be done (at least, I don't immediate see a way - please speak up if you can think of one).

christopher-henderson commented:
Howdy folks,

Between the log4j shenanigans and the holidays I myself took December largely off from the repo, however I'm hoping to push this feature over the finish line because I think it can really unlock a good number of doors for us with regard to ZLint's power and accuracy.

I do appreciate that 1.7k additions is a bit daunting, but I hope that knowing that more than 50% of that is just unit tests that are forced to be a bit redundant can help ease the apprehension.

Please do not hesitate if you would like more from me before attempting to tackle this review.

@cpu @zakird @sleevi

zakird commented:
I'm just beyond underwater this month, but can take a pass on January 5th.

cpu commented:
I'm taking a break from volunteer open source work for the short term and won't be available to review this PR.

zakird commented on v3/lints/cabf_br/lint_dsa_correct_order_in_subgroup_test.go:
```diff
@@ -41,7 +41,7 @@ func TestDSANotCorrectOrderSubgroup(t *testing.T) {
 	dsaKey.Y = pMinusOne
 	expected := lint.Error
-	out := test.TestLintCert("e_dsa_correct_order_in_subgroup", c)
+	out := test.TestLintCert("e_dsa_correct_order_in_subgroup", c, lint.NewEmptyConfig())
```
Is there a reason to pass in an EmptyConfig here every time instead of the Default Config?

christopher-henderson replied:
Hmmm it's because there is actually merge logic at play here. As in, we take a configuration from the frontend, we take the defaults given by the lint, and merge them together with preference give to the frontend. So if an empty confguration is given by the frontend then the lint's default values are used.

zakird commented on README.md:
```diff
@@ -96,6 +96,12 @@ Example ZLint CLI usage:
+    zlint -config configFile.toml mycert.pem
 
 See `zlint -h` for all available command line options.
 
```
Do we need to update library usage to explain how to pass in a configuration? Do we also want someone to be able to programmatically provide configurations instead of requiring a toml based interface? It's a bit awkward to have a TOML based structure to this instead of TOML encoding an internal structure that the library expects. This could be solved by having Configuration{} object be initializable using data instead of toml.

christopher-henderson replied:
hmmm I admint that I completely forgot about using ZLint as a library.

I'll have to come back to this one, but at the end of the day it does all resolve down to Go structs. But honestly my answer to this depends on which layer the the library interface is exposed, which I'll get on top of.

christopher-henderson commented on README.md:
```diff
@@ -96,6 +96,12 @@ Example ZLint CLI usage:
 	zlint -excludeSources=ETSI_ESI mycert.pem
+    
+    echo "Receive a copy of the full (default) configuration for all configurable lints"
+    zlint -exampleConfing
```
So the idea here was there is likely only going to be a small subset of lints that are actually configurable (of course, starting with 0). This makes the following questions difficult to answer - `Which lints are configurable? Is there documentation for these fields? What are their default values? If I am lint author, how do I convey the information to the user?`.

To date, the project has done a pretty good job of making it such that all a new lint has to do is add one new file, one new test file, and any necessary PEM data for tests. That is, everything is isolated making new lints rather easy for people unfamiliar to the codebase to contribute. However, if we naively asked that any new lint configurations kindly update the documentation on the current configuration, then that opens up a world of bookkeeping that can be quite lossy and error prone.

So what I've done here is make the following contract with lint authors...

> The code is going to require that you give me a default anyways, and our TOML framework lets you annotate your struct with comments, so whatever you give me as being your default configuration I will serialize into TOML and show to the end user on your behalf (comments and all).

In this way, new lints don't have to update markdown anywhere or otherwise leave the safety of their one file...

```go
type MyConfig struct {
   TheBestSongInTheWorld string `coment:"Totally your opinion"`
}

func (m (*MyLinT) Configure() interface{} {
    return &MyConfig{TheBestSongInTheWorld: "But I think it's Triangle Man by They Might be Giants"}
}
```

....and automagically that shows up in this "example" default configuration.

```toml
[MyLint]
# Totally your opinion
TheBestSongInTheWorld = "But I think it's Triangle Man by They Might be Giants"
```

christopher-henderson replied:
```suggestion
    zlint -exampleConfig
```

christopher-henderson commented on v3/cmd/zlint/main.go:
```diff
@@ -95,6 +99,16 @@ func main() {
+	if exampleConfig {
+		b, err := registry.DefaultConfiguration()
+		if err != nil {
+			fmt.Println(err)
```
```suggestion
			log.Fatalf("a critical error occurred while generating a configuration file, %s", err)
```

christopher-henderson commented on v3/cmd/zlint/main.go:
```diff
@@ -95,6 +99,16 @@ func main() {
+		b, err := registry.DefaultConfiguration()
+		if err != nil {
+			log.Fatalf("a critical error occurred while generating a configuration file, %s", err)
+			os.Exit(1)
```
```suggestion
```

christopher-henderson commented on v3/lint/base.go:
```diff
@@ -87,14 +93,33 @@ func (l *Lint) CheckEffective(c *x509.Certificate) bool {
+
+func (l *Lint) execute(lint LintInterface, cert *x509.Certificate, config Configuration) *LintResult {
+	switch configurable := lint.(type) {
+	case Configurable:
```
Good...point. Don't know why I did that, honestly.

aarongable commented on v3/lint/global_configurations.go:
```diff
@@ -0,0 +1,155 @@
+// [RFC5280Config]
+// ...
+// ...
+type RFC5280Config struct{}
```
Is the idea for these pre-defined config structs to gain various useful fields as part of this change, or is this PR only going to add the scaffolding while individual configurable fields will be added to these later?

christopher-henderson replied:
The later (if I understood you correctly).

So the idea for these structs is to have a ready-to-go facility for when someone (inevitably) finds a configuration that is valid for a broad classification of lints. What I've done here is, as you say, set up the scaffolding such that when that day does eventually arrive it is quite clear how it would be accomplished.

aarongable commented on v3/lint/registration.go:
```diff
@@ -87,6 +92,8 @@ type Registry interface {
 	// WriteJSON writes a description of each registered lint as
 	// a JSON object, one object per line, to the provided writer.
 	WriteJSON(w io.Writer)
+	SetConfiguration(config Configuration)
+	GetConfiguration() Configuration
```
As a library consumer of zlint, I just want to say that these new interface methods will work great for us. We [build a registry](https://github.com/letsencrypt/boulder/blob/7dcbf69536eba2d7133067c56e959873f00b4d0b/linter/linter.go#L148-L163) with our own configurable set of skip-lints; adding one more step to load and apply a config from a TOML file will be very easy.

aarongable commented on v3/lint/registration.go:
```diff
@@ -292,6 +301,63 @@ func (r *registryImpl) WriteJSON(w io.Writer) {
 	}
 }
 
+func (r *registryImpl) SetConfiguration(ctx Configuration) {
+	r.configuration = ctx
```
nit: save `ctx` for `context.Context` just to avoid confusing readers that are used to that custom.
```suggestion
func (r *registryImpl) SetConfiguration(cfg Configuration) {
	r.configuration = cfg
```

aarongable commented on v3/lint/registration.go:
```diff
@@ -313,7 +379,7 @@ var globalRegistry = NewRegistry()
 //
 // IMPORTANT: RegisterLint will panic if given a nil lint, or a lint with a nil
-// Lint pointer, or if the lint's Initialize function errors, or if the lint
+// Lint pointer, or if the lint's Initialize function errors, or if the lint`
```
typo: stray backtick
```suggestion
// Lint pointer, or if the lint's Initialize function errors, or if the lint
```

aarongable commented on v3/test/configuration_test_framework_test.go:
```diff
@@ -0,0 +1,203 @@
+`
+	out := TestLintWithConfig("e_ca_common_name_missing2", inputPath, config)
+	if out.Details != "liedershousen" {
+		panic("noooooooooooo")
```
Why panic instead of `t.Errorf(...)` here? (And a few other places below.)

christopher-henderson replied:
Simply some first-pass yolo test code that I missed when sprucing everything up :smiley:

aarongable commented on v3/test/helpers.go:
```diff
@@ -35,7 +39,15 @@ import (
+	return TestLintWithConfig(lintName, testCertFilename, "")
+}
+
+func TestLintWithConfig(lintName string, testCertFilename string, context string) *lint.LintResult {
+	ctx, err := lint.NewConfigFromString(context)
+	if err != nil {
+		panic(err)
+	}
```
I think the more idiomatic thing to do would be to take `t *testing.T` as the first argument, declare this a `t.Helper()`, and then use `t.Fatal()` instead of directly `panic`ing.

But I recognize that that's not the pattern used by the other helpers in this file, so it's probably better to do it this way for now.

christopher-henderson replied:
Right, I wouldn't be opposed to that were it not for the massive pile of lint tests that have already been written using this interface (a perfunctory glance shows ~550) :sob:

aarongable reviewed (commented):
Just dropping by to say that I'm very excited for this.

The use-case I have in mind right now is writing a lint which checks that the SCTs embedded in a cert abide by the new Chrome "two different operators" CT Policy. Doing that lint properly requires having a list of which logs are run by which operators, which means loading https://www.gstatic.com/ct/log_list/v3/log_list.json from disk, which means having a config which points to that file on disk.

christopher-henderson commented on README.md:
```diff
@@ -96,6 +96,12 @@ Example ZLint CLI usage:
+    zlint -exampleConfig
+
+    echo "Lint mycert.pem using a custom configuration for any configurable lints"
+    zlint -config configFile.toml mycert.pem
```
Tricksy four spaces instead of tabs, indeed ;)

christopher-henderson commented on v3/lint/configuration.go:
```diff
@@ -0,0 +1,248 @@
+// exists then that higher scoped configuration will be copied into the value
+// held by the provided interface{}.
+//
+// This procedure is recursive.
```
You are correct that there is a caching opportunity here, however I was apprehensive when building this in providing lints with shared data for which I do not believe I can reasonably control the mutability.

```go
struct A {
    Config *RFC5280Config
}

func (a *A) Execute(cert *x509.Certificate) {
    A.Config.Field = "now something else entirely"
}

struct A {
    Config *RFC5280Config
}

func (b *B) Execute(cert *x509.Certificate) {
    // I sure hope that no one modified my config before I ran because that would be super sneaky
}
```

I suppose that for the common configurations I could do a series of `getters`, although I cannot do that in general for all configurations as I cannot know their beforehand.

However, a quick measurement shows that this is not much a problem (at least not at the moment).

```go
func BenchmarkConfig(b *testing.B) {
	config, err := lint.NewConfigFromString(`
[e_ca_common_name_missing2]
BeerHall = "liedershousenssss"
`)
	if err != nil {
		b.Fatal(err)
	}
	lint := NewCaCommonNameMissing()
	for i := 0; i < b.N; i++ {
		err = config.Configure(lint, "e_ca_common_name_missing2")
		if err != nil {
			b.Fatal(err)
		}
	}
}
```

...yields...

```
goos: linux
goarch: amd64
pkg: github.com/zmap/zlint/v3/test
cpu: Intel(R) Core(TM) i7-6700K CPU @ 4.00GHz
BenchmarkConfig
BenchmarkConfig-8   	  931068	      1250 ns/op
PASS
```

And indeed, simply tight looping this reconstruction 100k times take about 0.13s.

christopher-henderson commented:
Thank you @zakird and @aarongable for your patience - other low hanging fruit kept presenting itself and it was just too tempting to go for the small wins too often.

@aarongable thank you so much for hopping in here without even being asked to! It is very much appreciated to receive community interest and support, and I have been very interested in receiving feedback from consumers on the usability if this framework as well as candidate usecases.

@zakird I looked into how this would all work from the perspective of a library usage and thankfully not much had to be done beyond a bug fix, some extra tests, and the following documentation.

### From the README
Library Usage
-------------
...
...
...

To lint a certificate in the presence of a particular configuration file, you must first construct the configuration and then make a call to `SetConfiguration` in the `Registry` interface.

A `Configuration` may be constructed using any of the following functions:

* `lint.NewConfig(r io.Reader) (Configuration, error)`
* `lint.NewConfigFromFile(config string) (Configuration, error)`
* `lint.NewConfigFromString(config string) (Configuration, error)`

The contents of the input to all three constructors must be a valid TOML document.

```go
import (
	"github.com/zmap/zcrypto/x509"
	"github.com/zmap/zlint/v3"
)

var certDER []byte = ...
parsed, err := x509.ParseCertificate(certDER)
if err != nil {
	// If x509.ParseCertificate fails, the certificate is too broken to lint.
	// This should be treated as ZLint rejecting the certificate
	log.Fatal("unable to parse certificate:", err)
}
configuration, err := lint.NewConfigFromString(`
        [some_configurable_lint]
        IsWebPki = true
        NumIterations = 42
        
        [some_configurable_lint.AnySubMapping]
        something = "else"
        anything = "at all"
`)
if err != nil {
	log.Fatal("unable to parse configuration:", err)
}
lint.GlobalRegistry().SetConfigutration(configuration)
zlintResultSet := zlint.LintCertificate(parsed)
```

aarongable commented on v3/cmd/zlint/main.go:
```diff
@@ -96,6 +100,15 @@ func main() {
+		if err != nil {
+			log.Fatalf("a critical error occurred while generating a configuration file, %s", err)
+		}
+		fmt.Printf("%s\n", string(b))
```
nit: I believe `fmt.Println(string(b))` would be equivalent here.

aarongable commented on README.md:
```diff
@@ -149,6 +155,45 @@ if err != nil {
+A `Configuration` may be constructed using any of the following functions:
+
+* `lint.NewConfig(r io.Reader) (Configuration, error)`
+* `lint.NewConfigFromFile(config string) (Configuration, error)`
```
nit: The argument to this function is called `path`, not `config`, in the actual function definition.

aarongable commented on v3/cmd/zlint/main.go:
```diff
@@ -203,6 +216,11 @@ func trimmedList(raw string) []string {
 // use.
 //nolint:cyclop
 func setLints() (lint.Registry, error) {
+	config, err := lint.NewConfigFromFile(config)
```
nit: this kind of shadowing (here, `config` is now of type `Configuration`, instead of type `string`) can be difficult to read. I'd use a different variable name here to avoid any chance of confusion.

aarongable commented on v3/test/helpers.go:
```diff
@@ -35,7 +39,15 @@ import (
+	return TestLintWithConfig(lintName, testCertFilename, "")
+}
+
+func TestLintWithConfig(lintName string, testCertFilename string, context string) *lint.LintResult {
+	ctx, err := lint.NewConfigFromString(context)
```
nit: calling it a `context` and naming the variable `ctx` is likely to be confusing to people who tightly associate that name with `context.Context`. I'd rename these `configStr` and `config`, or `config` and `cfg`, or something like that.

The same also applies a couple times below where the name `ctx` is used for a `lint.Configuration`.

aarongable reviewed (commented):
All the core logic code looks good to me, just a few nits around the documentation and naming!

christopher-henderson commented:
Howdy @aarongable, could you take one more perfunctory glance at this after I addresses some of the nits (I honestly have no idea why GH won't let me cycle your review status...). I think we're just about ready to get this change in in preparation for a minor version bump :smile:
